src/corner_kick/src/types/spritesheet.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * This file specifies types for spritesheet generation
+ */
+
+/**
+ * The various types of shapes that can be drawn in a spritesheet
+ */
+export enum ShapeType {
+    RECT = 'rect',
+    ELLIPSE = 'ellipse',
+    LINE = 'line',
+    ARC = 'arc',
+    POLY = 'poly',
+}
+
+/**
+ * Description object for spritesheet. Used by
+ * `SpritesheetManager` to generate a spritesheet.
+ */
+export interface ISpritesheet {
+    /**
+     * The frames contained in this spritesheet
+     */
+    frames: IFrame[];
+    /**
+     * The size of the spritesheet, make sure all the frames are contained
+     * inside these dimensions
+     */
+    dimensions: {
+        h: number;
+        w: number;
+    };
+}
+
+/**
+ * Describes a single frame in a spritesheet, which may contain many shapes
+ */
+export interface IFrame {
+    /**
+     * The dimension of the frame, any shape drawn outside of these dimensions
+     * will be cropped
+     */
+    dimensions: { x: number; y: number; w: number; h: number };
+    /**
+     * The shapes of this frame, drawn relative to the frame top-left corner
+     */
+    shapes: Array<IRectShape | IEllipseShape | IArcShape | ILineShape | IPolyShape>;
```

Comment:
Could we just have 
``` suggestion
    shapes: Array<IShape>;
```
instead?

Reply:
No, this is needed to have type checking specific to each type of shape that can be drawn

Reply:
I'm a bit confused about this. Isn't the point of using an interface like `IShape` (in an object-oriented sense) that we know that anything we pass in can only be treated as an `IShape` (ie. you can't call members that are not part of the `IShape`)? This might be a discussion better had in person or over slack.

Reply:
The thought was to use Typescript typechecking. But I can simplify this if you feel so inclined

Reply:
Fixed

Reply:
Sorry, I'm not objecting to this directly, I'm just confused about why we need to do this to allow Typescript to do it's typechecking? Again, this might be a better discussion over slack or in person.

Reply:
Let me try to explain it. We use a type union for all shape types because we expect one of five shapes, but we are not sure which one. A type guard (a switch block on the `shape.type` field) allows us to identify the shape provided and the union resolves to one specific shape type.

https://codingblast.com/typescript-union-types-type-guards-type-aliases/

Reply:
Regardless, I removed this approach as it may cause confusion

Reply:
Why wouldn't we have just used inheritance here? I could imagine that if we had an `IShape` object with a `drawOnCanvas` method, we could just accept that, and typescript would reasonably be able to check that the types all matched up? We could then just call the `drawOnCanvas` method with anything that's passed in, regardless of what specialisation of `IShape` we had?

Reply:
This is an interface, aka a JSON object. These types of objects cannot have methods if they are meant to be serializable. They just contain shape data. We need an external function to identify the type of shape and call the appropriate draw method, as seen in `draw.ts`

Reply:
Interesting. I'm chill with this, but would like to chat more in person to really understand it.